### ISSUE TITLE
False positive in cast_ptr_alignment

### ISSUE THREAD

**Issue by djugei** [rust-lang/rust-clippy#5618]
if i cast a highly-aligned pointer to u8 to a mid-aligned pointer (in my case its 4096 -> 1 -> 8) clippy errors. this is wrong. there are many cases where casting from u8 is acceptable alignment-wise, including when you have manually checked that the alignment works out, statically or dynamically.

**Comment by flip1995** [comment]
Do you have a code example for this?

**Comment by djugei** [comment]
sure thing: https://github.com/djugei/chunk-list

**Comment by flip1995** [comment]
So a minimal example would like this:

```rust
let _ = (ptr as *const u8) as *const u64;
```

This may be ok for `align_of(ptr) >= 8byte`, which could be fixed for the example above.

> including when you have manually checked that the alignment works out, statically or dynamically.

In (special) cases where you disagree with a lint, or have manually checked that a lint is not applicable, you should use `allow(clippy::lint)` to signal that the following code is really intended. 

IMO casts from u8 pointer is not a FP in general, but a special case where a cast to a more strictly aligned pointer is fine ➡️ `allow`.

**Comment by djugei** [comment]
> why would you cast to u8 first
because i want to add an offset in bytes.

i honestly think this lint should be completely removed, clippy is not equiped to deal with this situation.
you can't really guess at the validity of pointer casts from a simple syntax analysis. i feel like this job is better suited for miri.

im reasonably sure that this lint will also error on the standard library, for example in [slice](https://doc.rust-lang.org/src/core/slice/mod.rs.html#2487-2510)

**Comment by flip1995** [comment]
> you can't really guess at the validity of pointer casts from a simple syntax analysis

That's a fair point. I don't think we should remove it completely, but rather move it to pedantic

**Comment by djugei** [comment]
i am not sure what further discussion is needed, i will open a pr.

**Comment by flip1995** [comment]
Thanks! I marked it as need-discussion, to have it over for some more opinions. But this was 10 days ago, so I think moving this is fine.
